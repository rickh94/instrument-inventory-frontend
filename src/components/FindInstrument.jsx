import React, { useState, useContext } from 'react'
import PropTypes from 'prop-types'
import { makeStyles } from '@material-ui/styles'
import { withRouter } from 'react-router-dom'

import { lastButton } from '../globalStyles'
import { API } from 'aws-amplify'
import { LoadingHeader, ScannerField } from '.'
import { Button, FormGroup } from '@material-ui/core'
import { HelpersContext } from '../contexts'

export const getPath = input =>
  input.match(/\w?\d+-\d+/) ? 'search/number' : 'search/assigned-history'

const useStyles = makeStyles(theme => ({
  lastButton,
  firstButton: { marginRight: theme.spacing(-1) },
}))

export const FindInstrument = ({ showMultipleResults, history }) => {
  const { showAlert } = useContext(HelpersContext)
  const [isLoading, setLoading] = useState(false)
  const [searchTerm, setSearchTerm] = useState('')
  const [error, setError] = useState('')
  const classes = useStyles()

  const onSubmit = async event => {
    if (event) {
      event.preventDefault()
    }

    if (!searchTerm) {
      setError('Please enter a search term')
      return
    }

    await getInstrument(searchTerm)
  }

  const getInstrument = async term => {
    setLoading(true)
    const path = getPath(term)

    try {
      const response = await API.post('instrument-inventory', path, {
        body: { term },
      })

      if (response.length == 1) {
        showAlert('Instrument found')
        showMultipleResults([])
        history.push(`/instrument/${response[0].id}`)
      } else {
        showMultipleResults(response)
      }
    } catch (err) {
      if (err.response) {
        showAlert(`Error: ${err.response.data}`)
      } else {
        showAlert(`Error: ${err}`)
      }
    }
    setLoading(false)
  }

  return (
    <>
      <LoadingHeader isLoading={isLoading} title="Find an Instrument" />
      <form onSubmit={onSubmit}>
        <ScannerField
          value={searchTerm}
          setValue={setSearchTerm}
          error={error}
          label="Name or Number"
          submitCallback={getInstrument}
        />
        <FormGroup row>
          <Button
            onClick={() => setSearchTerm('')}
            className={classes.lastButton}
            data-testid="clear-button"
          >
            Clear
          </Button>
          <Button
            type="submit"
            color="primary"
            disabled={searchTerm === ''}
            data-testid="submit-button"
            className={classes.firstButton}
          >
            Submit
          </Button>
        </FormGroup>
      </form>
    </>
  )
}

FindInstrument.propTypes = {
  showMultipleResults: PropTypes.func.isRequired,
  history: PropTypes.object.isRequired,
}

export default withRouter(FindInstrument)
