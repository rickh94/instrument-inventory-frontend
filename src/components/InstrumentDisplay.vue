<template>
  <v-modal v-if="currentInstrument"
           @close="editing = false; clearCurrentInstrument()"
           :width-class="editing ? 'sm:max-w-xl' : 'sm:max-w-lg'">
    <div class="rounded w-full sm:max-w-2xl" v-if="editing">
      <v-instrument-form
        @editSuccess="editing = false"
        @cancel="editing = false"
        class="max-h-screen overflow-auto overflow-x-hidden rounded"></v-instrument-form>
    </div>
    <div v-else class="w-full">
      <div class="flex justify-between mb-4">
        <h6 class="text-xl font-bold text-gray-900">
          <span v-if="currentInstrument.size !== 'N/A'">
           {{ currentInstrument.size }}
          </span>
          {{ currentInstrument.type }} {{
            currentInstrument.number }}
        </h6>
        <button class="appearance-none text-white bg-red-600 px-1 rounded shadow hover:bg-red-800 hover:shadow-lg text-sm"
                @click="clearCurrentInstrument" title="Close Instrument Display">
          <svg xmlns="http://www.w3.org/2000/svg" class="h-5 w-5" viewBox="0 0 20 20" fill="currentColor">
            <path fill-rule="evenodd"
                  d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
                  clip-rule="evenodd" />
          </svg>
        </button>
      </div>
      <div>
        <span class="font-bold text-gray-700 mb-2">Size: </span>{{ currentInstrument.size }}
      </div>
      <div>
        <span class="font-bold text-gray-700 mb-2">Location: </span>{{ currentInstrument.location }}
      </div>
      <div>
        <span class="font-bold text-gray-700 mb-2">Assigned To: </span>{{ currentInstrument.assignedTo }}
      </div>
      <div>
        <span class="font-bold text-gray-700 mb-2">Notes: </span>{{
          currentInstrument.conditionNotes }}
      </div>
      <div>
        <span class="font-bold text-gray-700 mb-2">Condition: </span>{{ currentInstrument.condition }}
      </div>
      <div>
        <span class="font-bold text-gray-700 mb-2">Quality: </span>{{ currentInstrument.quality }}
      </div>
      <div>
        <span class="font-bold text-gray-700 mb-2">Gifted To Student: </span>{{ currentInstrument.gifted ?
        "Yes" : "No" }}
      </div>
      <div v-if="currentInstrument.history">
        <span class="font-bold text-gray-700 mb-2">History: </span>{{ currentInstrument.history.join(", ")
        }}
      </div>
      <div class="font-bold text-gray-700" v-else>
        No Instrument History
      </div>
      <div>
        <span class="font-bold text-gray-700 mb-2">Archived</span> {{ currentInstrument.archived ? "Yes" : "No" }}
      </div>
      <v-loading-buttons loader-class="w-80 mr-2" :loading="loading" container-class="mt-2 grid grid-cols-2 sm:grid-cols-4 gap-y-2">
        <v-edit-button v-if="isAdmin" @click="edit" center />
        <router-link
          v-if="$route.path !== '/sign-out'"
          to="/sign-out"
          class="inline-flex items-center justify-center font-bold mx-1 appearance-none bg-blue-600 text-white px-3 py-1 shadow rounded hover:bg-blue-800 hover:shadow-lg">
          <span class="inline-flex items-center">
            <svg xmlns="http://www.w3.org/2000/svg" class="h-5 w-5 mr-1" viewBox="0 0 20 20" fill="currentColor"> <path d="M9 2a1 1 0 000 2h2a1 1 0 100-2H9z" />
              <path fill-rule="evenodd"
                    d="M4 5a2 2 0 012-2 3 3 0 003 3h2a3 3 0 003-3 2 2 0 012 2v11a2 2 0 01-2 2H6a2 2 0 01-2-2V5zm3 4a1 1 0 000 2h.01a1 1 0 100-2H7zm3 0a1 1 0 000 2h3a1 1 0 100-2h-3zm-3 4a1 1 0 100 2h.01a1 1 0 100-2H7zm3 0a1 1 0 100 2h3a1 1 0 100-2h-3z"
                    clip-rule="evenodd" />
            </svg>
            Assign
          </span>
        </router-link>
        <v-retrieve-button @click="retrieve" center />
        <button v-if="isAdmin" @click="toggleArchived"
                class="inline-flex items-center justify-center font-bold mx-1 appearance-none bg-pink-600 text-white px-4 py-1 shadow rounded hover:bg-pink-800 hover:shadow-lg">
          <span class="inline-flex items-center">
            <svg xmlns="http://www.w3.org/2000/svg"
                 class="h-6 w-6 mr-1"
                 fill="none"
                 viewBox="0 0 24 24"
                 stroke="currentColor"
                 v-if="currentInstrument.archived">
              <path stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    d="M5 8h14M5 8a2 2 0 110-4h14a2 2 0 110 4M5 8v10a2 2 0 002 2h10a2 2 0 002-2V8m-9 4h4" />
            </svg>
            <svg xmlns="http://www.w3.org/2000/svg" class="h-5 w-5 mr-1" viewBox="0 0 20 20" fill="currentColor" v-else>
              <path d="M4 3a2 2 0 100 4h12a2 2 0 100-4H4z" />
              <path fill-rule="evenodd"
                    d="M3 8h14v7a2 2 0 01-2 2H5a2 2 0 01-2-2V8zm5 3a1 1 0 011-1h2a1 1 0 110 2H9a1 1 0 01-1-1z"
                    clip-rule="evenodd" />
            </svg>
            {{ currentInstrument.archived ? "Un-archive" : "Archive" }}
          </span>
        </button>
      </v-loading-buttons>
<!--      <div class="mt-2 grid grid-cols-2 sm:grid-cols-4 gap-y-2" v-else>-->
<!--      </div>-->
    </div>
  </v-modal>
</template>
<script>
import { mapMutations, mapState } from "vuex";
import VModal from "@/components/UI/VModal";
import { API } from "aws-amplify";
import checkAdmin from "@/mixins/checkAdmin";
import VRetrieveButton from "@/components/UI/buttons/VRetrieveButton";
import VEditButton from "@/components/UI/buttons/VEditButton";
import VLoadingButtons from "@/components/UI/VLoadingButtons";

export default {
  name: "InstrumentDisplay",
  mixins: [checkAdmin],
  data() {
    return {
      editing: false,
      loading: false,
    };
  },
  components: {
    VLoadingButtons,
    VEditButton,
    VRetrieveButton,
    VInstrumentForm: () => import("@/components/createComponents/VInstrumentForm"),
    VModal,
  },
  methods: {
    ...mapMutations(["clearCurrentInstrument", "setCurrentInstrument", "updateCurrentInstrument", "clearNewInstrumentNumber"]),
    async toggleArchived() {
      const wasArchived = this.currentInstrument.archived;
      try {
        this.loading = true;
        const response = await API.put("instrument-inventory", `instruments/${this.currentInstrument.id}`, {
          body: {
            ...this.currentInstrument,
            archived: !this.currentInstrument.archived,
          },
        });
        this.loading = false;
        if (wasArchived) {
          this.$toasted.info(`Instrument ${response.item.number} un-archived`, { duration: 2000 });
        } else {
          this.$toasted.info(`Instrument ${response.item.number} archived`, { duration: 2000 });
        }
        this.updateCurrentInstrument(response.item);
      } catch (e) {
        this.loading = false;
        this.$toasted.error(`Error: ${e.response.data}`, { duration: 2000 });
      }
    },
    async retrieve() {
      this.loading = true;
      try {
        const response = await API.post("instrument-inventory", "retrieve-single",
          { body: { number: this.currentInstrument.number } });
        console.log(response);
        this.updateCurrentInstrument(response.item);
        this.$toasted.info(response.message, { duration: 2000 });
        this.loading = false;
      } catch (e) {
        this.$toasted.error(e.response.data, { duration: 2000 });
        console.log(e);
        this.loading = false;
      }

    },
    async handleEditSuccess() {
      this.editing = false;
      this.$emit("instrumentUpdated", this.currentInstrument.id);
    },
    edit() {
      this.clearNewInstrumentNumber();
      this.editing = true;
    },
  },
  computed: mapState(["currentInstrument"]),
};
</script>
