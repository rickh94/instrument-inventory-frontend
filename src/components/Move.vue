<template>
  <div :class="flash ? 'bg-green-300' : 'bg-white'">
    <h4 class="text-gray-800 font-bold text-xl mb-1">Move Instruments</h4>
    <form @submit.prevent="onSubmit" autocomplete="off">
      <v-form-control label="Location">
        <v-autocomplete :completion-options="acOptions.locations" v-model="location" required></v-autocomplete>
      </v-form-control>
      <label for="number" class="text-sm text-gray-800 mb-2">Instrument Number</label>
      <div class="flex items-start justify-between mt-2 mb-1">
        <div class="border-b focus-within:border-purple-300 border-gray-800 w-24 flex-grow flex items-center space-between">
          <input type="text"
                 name="number"
                 id="number"
                 class="appearance-none bg-transparent border-none w-full text-gray-800 mr-1 py-1 leading-tight focus:outline-none"
                 v-model="currentNumber"
                 @keydown.enter.prevent=""
                 @keyup.enter.prevent="onAdd"
          >
          <button class="appearance-none" title="Scan Barcode" @click.prevent="scanner = true">
            <font-awesome-icon icon="barcode" class="mx-2"></font-awesome-icon>
          </button>
        </div>
        <button @click.prevent="onAdd"
                class="appearance-none rounded ml-4 text-white font-bold px-2 py-1 inline-flex items-center"
                :class="currentNumber.length === 0 ? 'bg-gray-600' : 'bg-purple-600 hover:shadow hover:bg-purple-800'"
                :disabled="currentNumber.length === 0"
        >
          <svg xmlns="http://www.w3.org/2000/svg" class="h-5 w-5 mr-1" viewBox="0 0 20 20" fill="currentColor">
            <path fill-rule="evenodd"
                  d="M10 18a8 8 0 100-16 8 8 0 000 16zm1-11a1 1 0 10-2 0v2H7a1 1 0 100 2h2v2a1 1 0 102 0v-2h2a1 1 0 100-2h-2V7z"
                  clip-rule="evenodd" />
          </svg>
          Add
        </button>
      </div>
      <v-scanner @detected="detected" v-if="scanner" @close="scanner = false"></v-scanner>
      <div class="my-2 max-w-xl mx-auto" v-show="this.instrumentNumbers.length > 0"><h4 class="text-xl font-bold">
        Instruments to Move</h4></div>
      <transition-group name="tag" tag="div" class="flex flex-wrap mx-auto max-w-xl" mode="out-in">
        <button v-for="instrument in instrumentNumbers"
                @click.prevent="remove(instrument)"
                :key="instrument"
                class="bg-purple-300 flex m-1 justify-between items-center rounded-full px-4 py-2 shadow-sm hover:shadow hover:bg-red-300"
                title="Click to remove">
          <span class="block mr-1">{{ instrument }}</span>
          <svg xmlns="http://www.w3.org/2000/svg" class="h-5 w-5" viewBox="0 0 20 20" fill="currentColor">
            <path fill-rule="evenodd"
                  d="M9 2a1 1 0 00-.894.553L7.382 4H4a1 1 0 000 2v10a2 2 0 002 2h8a2 2 0 002-2V6a1 1 0 100-2h-3.382l-.724-1.447A1 1 0 0011 2H9zM7 8a1 1 0 012 0v6a1 1 0 11-2 0V8zm5-1a1 1 0 00-1 1v6a1 1 0 102 0V8a1 1 0 00-1-1z"
                  clip-rule="evenodd" />
          </svg>
        </button>
      </transition-group>
      <div class="flex justify-end mt-4 mb-2" v-if="loading">
        <bar-loader class="w-56 mr-2" color="#7c3aed"></bar-loader>
      </div>
      <div v-else class="flex justify-end w-full mt-5">
        <button class="bg-red-600 mx-2 px-3 text-white py-2 shadow hover:bg-red-800 hover:shadow-lg rounded font-bold inline-flex items-center"
                type="reset"
                @click.prevent="instrumentNumbers = []">
          <svg xmlns="http://www.w3.org/2000/svg" class="h-5 w-5 mr-1" viewBox="0 0 20 20" fill="currentColor">
            <path fill-rule="evenodd"
                  d="M10 18a8 8 0 100-16 8 8 0 000 16zM7 9a1 1 0 000 2h6a1 1 0 100-2H7z"
                  clip-rule="evenodd" />
          </svg>
          Clear
        </button>
        <button class="bg-purple-600 ml-2 px-3 font-bold text-white py-2 shadow hover:bg-purple-800 hover:shadow-lg rounded inline-flex items-center"
                type="submit">
          <svg xmlns="http://www.w3.org/2000/svg"
               class="h-6 w-6 mr-1"
               fill="none"
               viewBox="0 0 24 24"
               stroke="currentColor">
            <path stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  d="M8 7h12m0 0l-4-4m4 4l-4 4m0 6H4m0 0l4 4m-4-4l4-4" />
          </svg>
          Move
        </button>
      </div>
    </form>
  </div>
</template>

<script lang="ts">
/* eslint-disable no-unused-vars */
import Component, { mixins } from "vue-class-component";
import { BarLoader } from "@saeris/vue-spinners";
import { GenericOutcome } from "@/util/commonTypes";
import VFormControl from "@/components/UI/VFormControl.vue";
import VAutocomplete from "@/components/UI/VAutocomplete.vue";
import acOptions from "@/mixins/acOptions";
import { moveMultiple } from "@/services/move";

enum AddOutcomes {
  Added = "ADDED",
  Invalid = "INVALID",
  Duplicate = "DUPLICATE"
}


@Component({
  components: { VAutocomplete, VFormControl, VScanner: () => import("@/components/UI/VScanner.vue"), BarLoader },
})
export default class Move extends mixins(acOptions) {
  instrumentNumbers: string[] = [];
  currentNumber = "";
  scanner = false;
  loading = false;
  flash = false;
  location = "";

  async created(): Promise<void> {
    await this.getACOptions();
  }

  async onSubmit(): Promise<void> {
    if (this.instrumentNumbers.length < 1) {
      this.$toasted.error("No Instruments to move", { duration: 2000 });
      return;
    }

    this.loading = true;
    const [outcome, result] = await moveMultiple(this.instrumentNumbers, this.location);
    this.loading = false;

    switch (outcome) {
      case GenericOutcome.Ok:
        this.handleUpdated(result.instrumentsUpdated);
        this.handleFailed(result.instrumentsFailed);
        break;
      case GenericOutcome.Err:
        this.$toasted.error(`Error: ${result}`, { duration: 2000 });
        console.error(result);
        break;
      default:
        this.$toasted.error("Something went wrong", { duration: 1000 });
    }
  }

  handleUpdated(updatedNumbers: string[]): void {
    if (updatedNumbers.length > 0) {
      this.$toasted.success(`Successfully moved ${updatedNumbers.join(", ")}`, { duration: 4000 });
    }
  }

  handleFailed(failedNumbers: string[]): void {
    if (failedNumbers.length > 0) {
      this.$toasted.success(`Failed to move ${failedNumbers.join(", ")}`, { duration: 4000 });
    }
  }

  onAdd(): void {
    switch (this.doAdd(this.currentNumber)) {
      case AddOutcomes.Added:
        this.currentNumber = "";
        break;
      case AddOutcomes.Duplicate:
        this.currentNumber = "";
        this.$toasted.info("Instrument already added", { duration: 900 });
        break;
      case AddOutcomes.Invalid:
        this.$toasted.error("Invalid Instrument number", { duration: 1000 });
        break;
      default:
        this.$toasted.error("Something went wrong", { duration: 1000 });
    }
  }

  detected(result: { codeResult: { code: string } }): void {
    switch (this.doAdd(result.codeResult.code)) {
      case AddOutcomes.Added:
        this.doFlash();
        break;
      case AddOutcomes.Duplicate:
      case AddOutcomes.Invalid:
        break;
      default:
        this.$toasted.error("Something went wrong", { duration: 1000 });
    }
  }

  remove(instrument: string): void {
    this.instrumentNumbers = this.instrumentNumbers.filter(item => item !== instrument);
  }

  doFlash(timeout = 300): void {
    this.flash = true;
    setTimeout(() => this.flash = false, timeout);
  }

  doAdd(number: string): AddOutcomes {
    if (!number.match(/\w*\d*-\d+/)) {
      return AddOutcomes.Invalid;
    }
    if (this.instrumentNumbers.includes(number)) {
      return AddOutcomes.Duplicate;
    }
    this.instrumentNumbers.push(number);
    return AddOutcomes.Added;
  }
}
</script>

<!--suppress CssUnusedSymbol -->
<style scoped>

.tag-enter-active {
  transition: opacity 0.5s;
}

.tag-enter {
  opacity: 0;
}

</style>
