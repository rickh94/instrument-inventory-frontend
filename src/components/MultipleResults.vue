<template>
  <div class="m-5 shadow p-3" v-show="$store.state.searchResults.length > 1">
    <h4 class="text-xl mb-1 font-bold text-gray-800">Multiple Instruments Found</h4>
    <div class="flex flex-col mx-2 mt-4 ">
      <div class="py-2 text-gray-800 flex justify-between border-b border-gray-600"
           v-for="instrument in $store.state.searchResults"
           :key="instrument.id">
        <button class="appearance-none hover:underline text-purple-600" @click="select(instrument)">{{ instrument.size }} {{
          instrument.type }} {{ instrument.number }}
        </button>
        <div>{{ instrument.assignedTo }}</div>
        <div v-if="instrument.history">{{ instrument.history.join(', ') }}</div>
        <div v-else>No instrument history</div>
      </div>
    </div>
  </div>
</template>
<script>
  import { mapMutations } from 'vuex'

  export default {
    name: 'MultipleResults',
    methods: {
      select(instrument) {
        this.setCurrentInstrument(instrument)
      },
      ...mapMutations(['setCurrentInstrument', 'clearCurrentInstrument']),
    },
  }
</script>
