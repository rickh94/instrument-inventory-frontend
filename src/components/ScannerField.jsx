import React, { useState } from 'react'
import PropTypes from 'prop-types'
import {
  FormControl,
  InputLabel,
  Button,
  Input,
  InputAdornment,
  FormHelperText,
} from '@material-ui/core'
import { Scanner, TooltipIconButton } from '.'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faBarcode } from '@fortawesome/free-solid-svg-icons'

const ScannerField = ({ error, value, setValue, label, submitCallback }) => {
  const autoScan = JSON.parse(localStorage.getItem('autoScan', 'false'))
  const [scanning, setScanning] = useState(autoScan)

  const onDetected = result => {
    if (result.codeResult.code !== value) {
      setScanning(false)
      if (submitCallback) {
        setValue(result.codeResult.code)
        submitCallback(result.codeResult.code)
      } else {
        setValue(result.codeResult.code)
      }
    }
  }

  return (
    <FormControl fullWidth error={error ? true : false}>
      <InputLabel data-testid="label" htmlFor="scanner-field">
        {label}
      </InputLabel>
      {scanning ? (
        <React.Fragment>
          <Scanner onDetected={onDetected} data-testid="scanner" />
          <Button onClick={() => setScanning(false)}>Stop Scanning</Button>
        </React.Fragment>
      ) : (
        <React.Fragment>
          <Input
            id="scanner-field"
            onChange={event => setValue(event.target.value)}
            aria-describedby="scanner-error"
            type="text"
            value={value}
            data-testid="input-field"
            required
            endAdornment={
              <InputAdornment position="end">
                <TooltipIconButton
                  onClick={() => setScanning(true)}
                  title="Scan Barcode"
                  data-testid="scan-button"
                >
                  <FontAwesomeIcon icon={faBarcode} />
                </TooltipIconButton>
              </InputAdornment>
            }
          />
          {error && <FormHelperText id="scanner-error">{error}</FormHelperText>}
        </React.Fragment>
      )}
    </FormControl>
  )
}

ScannerField.propTypes = {
  error: PropTypes.string,
  value: PropTypes.string.isRequired,
  setValue: PropTypes.func.isRequired,
  label: PropTypes.string.isRequired,
  submitCallback: PropTypes.func,
}

export default ScannerField
