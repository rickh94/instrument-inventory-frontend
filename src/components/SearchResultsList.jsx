import React from 'react'
import PropTypes from 'prop-types'
import { withRouter } from 'react-router-dom'

import { Table, TableRow, TableBody, TableCell, TableHead } from '@material-ui/core'
import { titleCase } from '../libs/titleCase'

export const ResultRowInternal = ({ item, history }) => (
  <TableRow
    hover
    onClick={() => history.push(`/instrument/${item.id}`)}
    component="tr"
    data-testid="result-row"
  >
    <TableCell component="td" data-testid="type-and-number">
      {titleCase(item.type)} {item.number}
    </TableCell>
    <TableCell component="td" data-testid="size">
      {item.size}
    </TableCell>
    <TableCell component="td" data-testid="assignedTo">
      {item.assignedTo || 'Not Assigned'}
    </TableCell>
    <TableCell component="td" data-testid="location">
      {item.location}
    </TableCell>
  </TableRow>
)

ResultRowInternal.propTypes = {
  item: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
}

const ResultRow = withRouter(ResultRowInternal)

const SearchResultsList = ({ results }) => (
  <Table component="table">
    <TableHead component="thead">
      <TableRow component="tr">
        <TableCell component="th">Instrument</TableCell>
        <TableCell component="th">Size</TableCell>
        <TableCell component="th">Assigned To</TableCell>
        <TableCell component="th">Location</TableCell>
      </TableRow>
    </TableHead>
    <TableBody component="tbody">
      {results.map(item => (
        <ResultRow item={item} key={item.id} />
      ))}
    </TableBody>
  </Table>
)

SearchResultsList.propTypes = {
  results: PropTypes.array.isRequired,
}

export default SearchResultsList
