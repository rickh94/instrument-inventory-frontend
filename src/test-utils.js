/* eslint-disable react/prop-types */
import React from 'react'
import { MuiThemeProvider, createMuiTheme, indigo } from '@material-ui/core'
import { HelpersContext, SchemaContext } from './contexts'
import { MemoryRouter } from 'react-router-dom'
import { render } from '@testing-library/react'

const theme = createMuiTheme({
  palette: {
    primary: indigo,
    secondary: {
      main: '#0A81D1',
    },
  },
})

export const showAlert = jest.fn()
export const setSearchResults = jest.fn()
export const setFilterResults = jest.fn()
export const searchResults = jest.fn()
export const filterResults = jest.fn()

export const TestTheme = ({ children }) => (
  <MuiThemeProvider theme={theme}>{children}</MuiThemeProvider>
)

export const TestHelpers = ({ children }) => (
  <HelpersContext.Provider
    value={{
      showAlert,
      setSearchResults,
      setFilterResults,
      searchResults,
      filterResults,
    }}
  >
    {children}
  </HelpersContext.Provider>
)

export const TestRouter = ({ children }) => <MemoryRouter>{children}</MemoryRouter>

export const TestSchema = ({ children, schema = { components: { schemas: {} } } }) => (
  <SchemaContext.Provider value={schema}>{children}</SchemaContext.Provider>
)

export const TestEverything = ({ children }) => (
  <TestRouter>
    <TestTheme>
      <TestHelpers>{children}</TestHelpers>
    </TestTheme>
  </TestRouter>
)

// eslint-disable-next-line no-undef
export const flushPromises = () => new Promise(setImmediate)

const customRender = (ui, options, schema = { components: { schemas: {} } }) =>
  render(<TestSchema schema={schema}>{ui}</TestSchema>, {
    wrapper: TestEverything,
    ...options,
  })

export * from '@testing-library/react'

export { customRender as render }
