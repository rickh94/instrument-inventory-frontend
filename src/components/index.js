import AppliedRoute from './AppliedRoute'
import AuthenticatedRoute from './AuthenticatedRoute'
import LoadingHeader from './LoadingHeader'
import LoadingScreen from './LoadingScreen'
import Nav from './Nav'
import RootPaper from './RootPaper'
import Scanner from './Scanner'
import SearchResultsList from './SearchResultsList'
import UnauthenticatedRoute from './UnauthenticatedRoute'
import InstrumentDisplay from './InstrumentDisplay'
import FindInstrument from './FindInstrument'
import TodoList from './TodoList'
import TooltipIconButton from './TooltipIconButton'
import InstrumentTable from './InstrumentTable'
import SchemaForm from './SchemaForm'
import ScannerField from './ScannerField'

export {
  AppliedRoute,
  AuthenticatedRoute,
  LoadingHeader,
  LoadingScreen,
  Nav,
  RootPaper,
  Scanner,
  SearchResultsList,
  UnauthenticatedRoute,
  InstrumentDisplay,
  FindInstrument,
  TodoList,
  TooltipIconButton,
  InstrumentTable,
  SchemaForm,
  ScannerField,
}
