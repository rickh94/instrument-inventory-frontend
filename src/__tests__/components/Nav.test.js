import React from 'react'
import { Nav, NavItemInternal } from '../../components/Nav'
import { render, fireEvent, cleanup } from '../../test-utils'
import 'jest-dom/extend-expect'

afterEach(cleanup)

describe('Nav', () => {
  it('matches snapshot', () => {
    const { container } = render(
      <Nav isAuthenticated={false} handleLogout={jest.fn()} />
    )

    expect(container).toMatchSnapshot()
  })

  it('opens drawer when button is clicked', () => {
    const { container, queryByTestId } = render(
      <Nav isAuthenticated={false} handleLogout={jest.fn()} />
    )

    expect(queryByTestId('nav-drawer')).not.toBeInTheDocument()

    fireEvent.click(container.querySelector('button[data-testid="open-drawer-button"]'))

    expect(queryByTestId('nav-drawer')).toBeInTheDocument()
  })

  describe('renders the correct button based on isAuthenticated', () => {
    it('renders login if isAuthenticated is false', () => {
      const { queryByTestId } = render(
        <Nav isAuthenticated={false} handleLogout={jest.fn()} />
      )
      expect(queryByTestId('login-button')).toBeTruthy()
      expect(queryByTestId('logout-button')).toBeFalsy()
    })
    it('renders logout if true', () => {
      const { queryByTestId } = render(
        <Nav isAuthenticated={true} handleLogout={jest.fn()} />
      )

      expect(queryByTestId('logout-button')).toBeTruthy()
      expect(queryByTestId('login-button')).toBeFalsy()
    })
  })

  it('calls handle logout when logout is clicked', async () => {
    const handleLogoutMock = jest.fn()
    const { container } = await render(
      <Nav isAuthenticated={true} handleLogout={handleLogoutMock} />
    )

    fireEvent.click(container.querySelector('button[data-testid="logout-button"]'))
    expect(handleLogoutMock).toHaveBeenCalled()
  })
})

describe('NavItemInternal', () => {
  it('matches snapshot', () => {
    const { container } = render(
      <NavItemInternal
        to="/"
        setDrawerOpen={jest.fn()}
        text="test"
        history={{ push: jest.fn() }}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('pushes history and closes drawer on click', () => {
    const setDrawerOpenMock = jest.fn()
    const to = '/test'
    const historyMock = { push: jest.fn() }
    const { container } = render(
      <NavItemInternal
        to={to}
        setDrawerOpen={setDrawerOpenMock}
        history={historyMock}
        text="test"
      />
    )

    fireEvent.click(container.querySelector('li[data-testid="item"]'))
    expect(setDrawerOpenMock).toHaveBeenCalledWith(false)
    expect(historyMock.push).toHaveBeenCalledWith(to)
  })

  it('renders icon and title from props', () => {
    const { getByText, getByTestId } = render(
      <NavItemInternal
        icon={<div data-testid="test-icon" />}
        to="/"
        setDrawerOpen={jest.fn()}
        text="test text"
        history={{ push: jest.fn() }}
      />
    )
    expect(getByText('test text')).toBeInTheDocument()
    expect(getByTestId('test-icon')).toBeInTheDocument()
  })
})
