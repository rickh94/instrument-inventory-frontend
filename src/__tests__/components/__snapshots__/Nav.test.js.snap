// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav matches snapshot 1`] = `
<div>
  <div
    class="makeStyles-root-1"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation4 MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary mui-fixed"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <button
          aria-label="Menu"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
          data-testid="open-drawer-button"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <h6
          class="MuiTypography-root makeStyles-grow-2 MuiTypography-h6 MuiTypography-colorInherit"
        >
          Instrument Inventory
        </h6>
        <a
          class="makeStyles-cleanLink-4"
          href="/login"
        >
          <button
            class="MuiButtonBase-root MuiButton-root makeStyles-barText-3 MuiButton-text MuiButton-colorInherit"
            data-testid="login-button"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Login
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </a>
      </div>
    </header>
  </div>
  <div
    class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
  />
  <div
    class="PrivateSwipeArea-root-135 PrivateSwipeArea-anchorLeft-136"
    style="width: 20px;"
  />
</div>
`;

exports[`NavItemInternal matches snapshot 1`] = `
<div>
  <li
    aria-disabled="false"
    class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
    data-testid="item"
    role="button"
    tabindex="0"
  >
    <div
      class="MuiListItemText-root"
    >
      <span
        class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
      >
        test
      </span>
    </div>
    <span
      class="MuiTouchRipple-root"
    />
  </li>
</div>
`;
